File: frontend/src/types/shareholding.ts
export interface QuarterlyHolding {
  quarter: string;
  promoter: string;
  fii: string;
  dii: string;
  government: string;
  public: string;
}

export interface MajorShareholder {
  rank: number;
  name: string;
  holdingPercent: number;
}

export interface ShareholdingResponse {
  quarterlyData: QuarterlyHolding[];
  latest: {
    promoter: string;
    fii: string;
    dii: string;
    government: string;
    public: string;
  };
  majorShareholders: MajorShareholder[];
  source: string;
}

File: backend/app/schemas/shareholding.py
from typing import Optional
from app.schemas.base import BaseSchema


class QuarterlyHolding(BaseSchema):
    quarter: str
    promoter: str
    fii: str
    dii: str
    government: str
    public: str


class LatestHolding(BaseSchema):
    promoter: str
    fii: str
    dii: str
    government: str
    public: str


class MajorShareholder(BaseSchema):
    rank: int
    name: str
    holding_percent: float


class ShareholdingResponse(BaseSchema):
    quarterly_data: list[QuarterlyHolding]
    latest: LatestHolding
    major_shareholders: list[MajorShareholder]
    source: str
